### PR TITLE
Finish EdgeCellTree class

### PR DESCRIPTION
**Pull request by JoerivanEngelen**
Merge main into ``edge`` branch

EDIT: The changeset with the original ``edge`` branch was larger than with the ``main``, so now comparing those.
Done the following compared to original ``edge`` branch:

- Merge ``main`` into ``edge``
- Fix ``EdgeCellTree`` initialize
- Fix ``locate_points_on_edge`` function and add tests
- Fix ``lines_intersect`` util and add test for it with edge cases
- Start working on a ``edges_on_edges`` query

### CODE REVIEW

**Comment by Huite** [comment]
All the main difficulties have already have been resolved with the `locate_edge` functionality for locating edges inside of polygons.

We need a helper to re-allocate the results arrays.
In this case, we just need to store 2 floats since an edge-edge intersection is a single point (rather than two points for edge-convex polygon), so `xy` is modified; we also need to call a different individual query function in the loop.

```python
@nb.njit(cache=True)
def locate_edge_edges_helper(
    edge_coords: FloatArray,
    tree: CellTreeData,
    offset: int,
) -> IntArray:
    n_edge = len(edge_coords)
    # Ensure the initial indices array isn't too small.
    n = max(n_edge, 256)
    indices = np.empty((n, 2, 2), dtype=IntDType)   # MODIFY THIS
    xy = np.empty((n, 2), dtype=FloatDType)

    total_count = 0
    indices_size = 0
    for edge_index in range(n_edge):
        a = as_point(edge_coords[edge_index, 0])
        b = as_point(edge_coords[edge_index, 1])

        while True:
            count, indices_size = locate_edge(
                a, b, tree, indices, xy, indices_size, edge_index + offset
            )  # MODIFY THIS
            if count != -1:
                break
            # Not enough capacity: grow capacity, discard partial work, retry.
            indices_size = total_count
            indices = grow(indices)
            xy = grow(xy)

        total_count += count

    return indices, xy, total_count
```

We also need to an updated function around this which takes care of the parallellisation:

```python
@nb.njit(cache=True, parallel=PARALLEL)
def locate_edges(box_coords: FloatArray, tree: CellTreeData, n_chunks: int):
    chunks = np.array_split(box_coords, n_chunks)
    offsets = np.zeros(n_chunks, dtype=IntDType)
    for i, chunk in enumerate(chunks[:-1]):
        offsets[i + 1] = offsets[i] + len(chunk)

    # Setup (dummy) typed lists for numba to store parallel results.
    indices = [np.empty((0, 2), dtype=IntDType) for _ in range(n_chunks)]
    intersections = [np.empty((0, 2, 2), dtype=FloatDType) for _ in range(n_chunks)]
    counts = np.empty(n_chunks, dtype=IntDType)
    for i in nb.prange(n_chunks):
        indices[i], intersections[i], counts[i] = locate_edges_helper(  # MODIFY THIS
            chunks[i], tree, offsets[i]
        )

    total_size = sum(counts)
    xy = np.empty((total_size, 2, 2), dtype=FloatDType)   # MODIFY THIS
    start = 0
    for i, size in enumerate(counts):
        end = start + size
        xy[start:end] = intersections[i][:size]
        start = end

    ii, jj = concatenate_indices(indices, counts)
    return ii, jj, xy
```

We can either duplicate the code, or work around it with a closure. A closure is preferred over passing a function as an argument because we can "bake" in the function entirely and numba will compile more efficiently (rather than having function lookup at runtime).



Everything is mostly general, since in our EdgeCellTree, we're still organized by bounding boxes.

The polygon specific logic in `locate_edge` is this:

```python
        if node["child"] == -1:
            for i in range(node["ptr"], node["ptr"] + node["size"]):
                bbox_index = tree.bb_indices[i]
                box = as_box(tree.bb_coords[bbox_index])
                box_intersect, _, _ = cohen_sutherland_line_box_clip(a, b, box)
                if box_intersect:
                    polygon = copy_vertices_into(
                        tree.vertices, tree.faces[bbox_index], polygon_work_array
                    )
                    face_intersects, c, d = cyrus_beck_line_polygon_clip(a, b, polygon)
                    if face_intersects:
                        # If insufficient capacity, exit.
                        if indices_size >= capacity:
                            return -1, indices_size
                        indices[indices_size, 0] = index
                        indices[indices_size, 1] = bbox_index
                        intersections[indices_size, 0, 0] = c.x
                        intersections[indices_size, 0, 1] = c.y
                        intersections[indices_size, 1, 0] = d.x
                        intersections[indices_size, 1, 1] = d.y
                        indices_size += 1
                        count += 1
            continue
```

In case of edge edge intersection, this becomes:

```python
        if node["child"] == -1:
            for i in range(node["ptr"], node["ptr"] + node["size"]):
                bbox_index = tree.bb_indices[i]
                tree_edge = tree.edges[bbox_index]
                p = to_point(tree_edge[0])
                q = to_point(tree_edge[1])
                edge_intersects, c = lines_intersect(a, b, p, q)   # MODIFY LINES_INTERSECT TO RETURN COMPUTED INTERSECTION
                if edge_face_intersects:
                    # If insufficient capacity, exit.
                    if indices_size >= capacity:
                        return -1, indices_size
                    indices[indices_size, 0] = index
                    indices[indices_size, 1] = bbox_index
                    intersections[indices_size, 0, 0] = c.x
                    intersections[indices_size, 0, 1] = c.y
                    indices_size += 1
                    count += 1
            continue
```

**Comment by Huite** [comment]
Okay, I spent a bit more time thinking about this. The primary issue is in a cartesian plane that edge-(convex) polygon intersection may have two intersections, and edge-edge only one. We can either make this fancy and try to take into account the shapes everywhere, or we just accept a useless duplicated intersection for the edge-edge case.

Then we use a closure to capture the edge or face specific intersection function:

```python
@nb.njit(inline="always")
def compute_edge_edge_intersect(tree, bbox_index, a, b) -> Tuple[bool, Point, Point]:
    tree_edge = tree.edges[bbox_index]
    p = to_point(tree_edge[0])
    q = to_point(tree_edge[1])
    intersects, c = lines_intersect(a, b, p, q)   # MODIFY LINES_INTERSECT TO RETURN COMPUTED INTERSECTION
    return intersects, c, c


@nb.njit(inline="always")
def compute_edge_face_intersect(tree, bbox_index, a, b) -> Tuple[bool, Point, Point]:
    box = as_box(tree.bb_coords[bbox_index])
    intersects, c, d = cohen_sutherland_line_box_clip(a, b, box)
    if intersects:
        polygon = copy_vertices_into(
            tree.vertices, tree.faces[bbox_index], polygon_work_array
        )
        intersects, c, d = cyrus_beck_line_polygon_clip(a, b, polygon)
    return intersects, c, d


def make_locate_edge(intersection_function):

    # Inlining this function drives compilation time through the roof. It's
    # probably also a rather bad idea, given its complexity: compared to looking
    # for either boxes or points, checking is more much complicated by involving
    # two intersection algorithms.
    @nb.njit(inline="never")
    def locate_edge(
        a: Point,
        b: Point,
        tree: CellTreeData,
        indices: IntArray,
        intersections: FloatArray,
        indices_size: int,
        index: int,
    ):
        # Check if the entire mesh intersects with the line segment at all
        tree_bbox = as_box(tree.bbox)
        tree_intersects, _, _ = cohen_sutherland_line_box_clip(a, b, tree_bbox)
        if not tree_intersects:
            return 0, indices_size

        V = to_vector(a, b)
        stack = allocate_stack()
        polygon_work_array = allocate_polygon()
        stack[0] = 0
        size = 1
        count = 0
        capacity = len(indices)

        while size > 0:
            node_index, size = pop(stack, size)
            node = tree.nodes[node_index]

            # Check if it's a leaf
            if node["child"] == -1:
                for i in range(node["ptr"], node["ptr"] + node["size"]):
                    bbox_index = tree.bb_indices[i]
                    intersects, c, d = intersection_function(tree, bbox_index, a, b)
                    if intersects:
                        # If insufficient capacity, exit.
                        if indices_size >= capacity:
                            return -1, indices_size
                        indices[indices_size, 0] = index
                        indices[indices_size, 1] = bbox_index
                        intersections[indices_size, 0, 0] = c.x
                        intersections[indices_size, 0, 1] = c.y
                        intersections[indices_size, 1, 0] = d.x
                        intersections[indices_size, 1, 1] = d.y
                        indices_size += 1
                        count += 1
                continue

            # Note, "x" is a placeholder for x, y here
            # Contrast with t, which is along vector
            node_dim = 1 if node["dim"] else 0
            dx = V[node_dim]
            if dx > 0.0:
                dx_left = node["Lmax"] - a[node_dim]
                dx_right = node["Rmin"] - b[node_dim]
            else:
                dx_left = node["Lmax"] - b[node_dim]
                dx_right = node["Rmin"] - a[node_dim]

            # Check how origin (a) and end (b) are located compared to box edges
            # (Lmax, Rmin). The box should be investigated if:
            # * the origin is left of Lmax (dx_left >= 0)
            # * the end is right of Rmin (dx_right <= 0)
            left = dx_left >= 0.0
            right = dx_right <= 0.0

            # Now find the intersection coordinates. These have to occur within in
            # the bounds of the vector. Note that if the line has no slope in this
            # dim (dx == 0), we cannot compute the intersection, and we have to
            # defer to the child nodes.
            if dx > 0.0:  # TODO: abs(dx) > EPISLON?
                if left:
                    t_left = dx_left / dx
                    left = t_left >= 0.0
                if right:
                    t_right = dx_right / dx
                    right = t_right <= 1.0
            elif dx < 0.0:
                if left:
                    t_left = 1.0 - (dx_left / dx)
                    left = t_left >= 0.0
                if right:
                    t_right = 1.0 - (dx_right / dx)
                    right = t_right <= 1.0
            # else dx == 0.0. In this case there's no info to extract from this
            # node. We'll fully defer to the children.
            left_child = node["child"]
            right_child = left_child + 1

            if left and right:
                stack, size = push(stack, left_child, size)
                stack, size = push(stack, right_child, size)
            elif left:
                stack, size = push(stack, left_child, size)
            elif right:
                stack, size = push(stack, right_child, size)

        return count, indices_size

    @nb.njit(cache=True)
    def locate_edges_helper(
        edge_coords: FloatArray,
        tree: CellTreeData,
        offset: int,
    ) -> IntArray:
        n_edge = len(edge_coords)
        # Ensure the initial indices array isn't too small.
        n = max(n_edge, 256)
        indices = np.empty((n, 2), dtype=IntDType)
        xy = np.empty((n, 2, 2), dtype=FloatDType)

        total_count = 0
        indices_size = 0
        for edge_index in range(n_edge):
            a = as_point(edge_coords[edge_index, 0])
            b = as_point(edge_coords[edge_index, 1])

            while True:
                count, indices_size = locate_edge(
                    a, b, tree, indices, xy, indices_size, edge_index + offset
                )
                if count != -1:
                    break
                # Not enough capacity: grow capacity, discard partial work, retry.
                indices_size = total_count
                indices = grow(indices)
                xy = grow(xy)

            total_count += count

        return indices, xy, total_count


    @nb.njit(cache=True, parallel=PARALLEL)
    def locate_edges(box_coords: FloatArray, tree: CellTreeData, n_chunks: int):
        chunks = np.array_split(box_coords, n_chunks)
        offsets = np.zeros(n_chunks, dtype=IntDType)
        for i, chunk in enumerate(chunks[:-1]):
            offsets[i + 1] = offsets[i] + len(chunk)

        # Setup (dummy) typed lists for numba to store parallel results.
        indices = [np.empty((0, 2), dtype=IntDType) for _ in range(n_chunks)]
        intersections = [np.empty((0, 2, 2), dtype=FloatDType) for _ in range(n_chunks)]
        counts = np.empty(n_chunks, dtype=IntDType)
        for i in nb.prange(n_chunks):
            indices[i], intersections[i], counts[i] = locate_edges_helper(
                chunks[i], tree, offsets[i]
            )

        total_size = sum(counts)
        xy = np.empty((total_size, 2, 2), dtype=FloatDType)
        start = 0
        for i, size in enumerate(counts):
            end = start + size
            xy[start:end] = intersections[i][:size]
            start = end

        ii, jj = concatenate_indices(indices, counts)
        return ii, jj, xy

    return locate_edges

# Use the closure to capture the specific intersection function
locate_edge_edges = make_locate_edge(intersection_function=compute_edge_edge_intersect)
locate_edge_faces = make_locate_edge(intersection_function=compute_edge_face_intersect)
```

And then, in the EdgeCelltree class, we just discard the duplicated points.

```python
        edge_coords = cast_edges(edge_coords)
        n_chunks = nb.get_num_threads()
        xy, indices = locate_edges(edge_coords, self.celltree_data, n_chunks)
        return np.ascontiguousarray(xy[:, 0]), indices
```

For testing, start by disabling Numba to see if it works; might be good to do a performance check (needs sizable inputs, e.g. at least a million).
I'd add a new module, add this logic, expose it in a separate celltree method, then compare with the other one. I expect identical results for what it's worth -- numba should not be hindered by this closure. It's a common practice in the numba codebase as well.

**Comment by Huite** [comment]
To maintain the testing, we could just return them in a dict:

```python
def make_locate_edge(intersection_function):
    # Your existing code for locate_edge, locate_edges_helper, locate_edges
    
    # Return a dictionary containing all functions you want to expose for testing
    return {
        "locate_edges": locate_edges,
        "locate_edges_helper": locate_edges_helper,
        "locate_edge": locate_edge
    }

# Use the closure to capture the specific intersection function
edge_edge_functions = make_locate_edge(intersection_function=compute_edge_edge_intersect)
edge_face_functions = make_locate_edge(intersection_function=compute_edge_face_intersect)

# Extract the main functions for normal use
locate_edge_edges = edge_edge_functions["locate_edges"]
locate_edge_faces = edge_face_functions["locate_edges"]

# Expose helper functions for testing
locate_edge_edge = edge_edge_functions["locate_edge"]
locate_edge_face = edge_face_functions["locate_edge"]
```

**Comment by JoerivanEngelen** [comment]
Thanks for the code implementation, with some minor modifications I could get it to work. I've added some unittests and have added an example for the documentation:

image

Could you take another look @Huite?

I'm still a bit doubting the names of the classes now: ``CellTree2d`` and ``EdgeCellTree2d``. Maybe we should rename ``CellTree2d`` to ``FaceCellTree2d`` (of course make sure the original namespace is preserved for backwards compatibility, which points to  ``FaceCellTree2d`` and throws a DeprecationWarning.) Or maybe it is better to rename``EdgeCellTree2d`` to ``CellTree1d``? What do you think?

File: docs/api.rst [hunk 1]
```diff
@@ -10,6 +10,11 @@ CellTree2d
 Changelog
 =========
 
+[Unreleased]
+------------
+- Add ``EdgeCellTree2d`` class to support 2D queries on 1D grids.
```
Comment (Huite) [inline docs/api.rst:15]: I wouldn't use the term "1D grid", because it's clearly 2D data. I'd suggest "1D network" or something -- it's what the UGRID conventions mention. It'll work for line strings and such as well, so I'd use the term "linear geometry" or "linear features".

File: examples/spatial_indexing_1d_grids.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
+"""
+Spatial indexing of 1D grids
```
Comment (Huite) [inline examples/spatial_indexing_1d_grids.py:2]: I'd state:

"Spatial indexing of 1D networks and linear geometry"

File: numba_celltree/demo.py [hunk 3]
```diff
@@ -121,3 +121,26 @@ def generate_disk(partitions: int, depth: int):
     y = np.sin(angles) * radii
     triang = matplotlib.tri.Triangulation(x, y)
     return np.column_stack((x, y)), triang.triangles
+
+
+def example_1d_grid():
```
Comment (Huite) [inline numba_celltree/demo.py:126]: 1d_network, maybe

**Review by Huite — APPROVED** [review]

**Comment by Huite** [comment]
> I'm still a bit doubting the names of the classes now: CellTree2d and EdgeCellTree2d. Maybe we should rename CellTree2d to FaceCellTree2d (of course make sure the original namespace is preserved for backwards compatibility, which points to FaceCellTree2d and throws a DeprecationWarning.) Or maybe it is better to renameEdgeCellTree2d to CellTree1d? What do you think?

I think CellTree2d is probably fine; in the future I'd like to add a CellTree3d for volumes. The Garth and Joy paper's is also about faces/volumes, so the Edge celltree is a modification.

For three dimensional linework we'd need an EdgeCellTree3d; I think the terminology works.
Maybe questionable is whether we'd then need a FaceCellTree3d for surfaces in 3D space -- at the moment, I can't think of a case where you'd have data in such a form, but maybe it'll come up if we work on DISU more. But we can always change names to FaceCellTree2D and deprecate later.

File: docs/api.rst (hunk 1 above)
Comment (JoerivanEngelen) [reply docs/api.rst]: Done!

File: examples/spatial_indexing_1d_grids.py (hunk 2 above)
Comment (JoerivanEngelen) [reply examples/spatial_indexing_1d_grids.py]: Done!

File: numba_celltree/demo.py (hunk 3 above)
Comment (JoerivanEngelen) [reply numba_celltree/demo.py]: Done
